lib/dRep/dRepImpl.py

```diff
@@ -0,0 +1,441 @@
+
+
+
+
+        #
+        ##
+        ###
+        ####
+        #####
+
+
+        def simple_return(msg, report_params_kwargs={}):
+            '''
+            Return with a simple message
+            '''
+            report_params = {
+                    'message': msg,
+                    'workspace_name': _globals.workspace_name,
+                    **report_params_kwargs
+                    }
+            report_info = _globals.kbr.create_extended_report(report_params)
+            output = {
+                'report_name': report_info['name'],
+                'report_ref': report_info['ref'],
+            }
+
+            return [output]
+
+
+
+        # 
+        ##
+        ### check unique UPAs
+        #### 
+        #####
+
+        if len(set(params['genomes_refs'])) < len(params['genomes_refs']):
+            msg = 'Removing duplicate input BinnedContigs'
+            logging.warning(msg)
+            _globals.warnings.append(msg) 
+            params['genomes_refs'] = list(dict.fromkeys(params['genomes_refs']))
+
+       
+
+
+
+        # 
+        ##
+        ### load BinnedContigs files (to scratch) and obj data
+        ####
+        #####
+        
+
+        #
+        if _globals.debug and params.get('skip_dl'):
+            bins_dir_name_l = params['bins_dir_name_l']
+            bins_dir_l = [os.path.join(_globals.shared_folder, bins_dir_name) for bins_dir_name in bins_dir_name_l]
+
+            for upa, bins_dir in zip(params['genomes_refs'], bins_dir_l):
+                BinnedContigs(upa, get_bins_dir='local', bins_dir=bins_dir)
+
+
+        # load from KBase 
+        else:
+
+            for upa in params['genomes_refs']:
+                bc = BinnedContigs(upa, get_bins_dir='download')
+                
+                if bc.is_empty():
+                    # never mention this bc again
+                    params['genomes_refs'].remove(upa)
+                    BinnedContigs.created_instances.remove(bc)
+                
+
+            if len(params['genomes_refs']) == 0:
+                msg = 'Sorry, please input at least one non-empty BinnedContigs'
+                raise ValueError(msg)
+
+
+
+        #
+        ##
+        ###
+        ####
+        #####
+
+        for bc in BinnedContigs.created_instances:
+            bc.calc_stats()
+
+
+
+        #
+        ##
+        ### pool
+        ####
+        #####
+        
+ 
+        binsPooled_dir = os.path.join(_globals.run_dir, 'binsPooled')
+        os.mkdir(binsPooled_dir)
+
+        for bc in BinnedContigs.created_instances:
+            bc.pool_into(binsPooled_dir)
+        
+        dprint("os.listdir(binsPooled_dir)", run={**locals(), **globals()})
+
+
+        #
+        ##
+        ### 
+        #### params
+        #####
+       
+
+        # flatten param groups (if any)
+
+        paramGrp_key_l = ['filtering', 'genome_comparison', 'clustering', 'scoring', 'warnings']
+
+        for paramGrp_key in paramGrp_key_l:
```

Comment:
The purpose of this section was very confusing to me at first. I would add a comment about it, saying something like the dRep CLI parameters are nested under several groups, and we need to extract and flatten out the parameters from the different groups.